### DISCUSSION TITLE
New API Questions & Observations

### DISCUSSION THREAD

**Discussion by zencyon** [oobabooga/text-generation-webui#4813] [Q&A]
Hey :)

Been using Ooba as textgen backend for running several discord bots for a long time, loving it <3
(trying to get the bots back online after the latest changes to the new API, openai ext)

I'm seeing some differences in the model outputs, maybe some settings have changed behind the scenes that alters their pre-prompt

Below are the new data variables I'm using for the openai api

```
data = {

        'mode': "chat",
        'character': character,
        'messages': history,

        #EDIT: 'your_name': username no longer used, is now 'name1' and ''name2'
        'name1': username,
        'name2': ai_name,

        'max_tokens': 200,

        'presence_penalty':0, 
        # from OpenAI docs: Number between -2.0 and 2.0. Positive values penalize new tokens based on whether they appear in the text so far, increasing the model's likelihood to talk about new topics.
        
        'frequency_penalty':0, 
        #number or null Optional Defaults to 0
        
        'temperature': 0.7, 
        # from OpenAI docs: Defaults to 1 What sampling temperature to use, between 0 and 2. Higher values like 0.8 will make the output more random, while lower values like 0.2 will make it more focused and deterministic. We generally recommend altering this or top_p but not both.
       
        'top_p': 1.0, 
        # from OpenAI docs: Defaults to 1 An alternative to sampling with temperature, called nucleus sampling, where the model considers the results of the tokens with top_p probability mass. So 0.1 means only the tokens comprising the top 10% probability mass are considered. We generally recommend altering this or temperature but not both.
}
```
**// user names**

~~has the api 'your_name' variable name changed or is it no longer being processed in Ooba via api?~~

**EDIT: after checking typing.py, 'your_name' is now replaced by 'name1' and 'name2'** 

**// roles**

history.append({"role": "user", "content": user_message}) 
history.append({"role": character, "content": ai_message}) - is the 'role' for the ai supposed to be set to "assistant" or character name?

- does it depend on the model what role to use "assistant", "###HUMAN" etc? 
(if yes, could be issue since we have long histories, switching models often)

some documentation would be good to clear things up

other things:
sometimes empty replies, didn't happen much if ever before (hard to pin down if model)

cheers!

**Comment by naticio** [comment]
how do you use the new api? I don't see it as part of extensions anymore and I even if I put it as a parameter at start --api my requests do not work!

**Reply by zencyon** [reply]
![image](https://github.com/oobabooga/text-generation-webui/assets/43258551/ce33c378-9333-4b08-83b1-f7384b07ba03)

I just enable these 3 and hit Apply
(i read the if you add --api --listen to the config it sometimes doesn't work the first time)

otherwise follow the example Ooba provided in the open ai api for setting up the response code
(you have to use the open ai extension now)

https://github.com/oobabooga/text-generation-webui/wiki/12-%E2%80%90-OpenAI-API#examples

**Comment by Madd0g** [comment]
I'm having issues with textgen API that I didn't have with the same models in ollama+litellm.

I was suspecting that templates get applied from the API calls too, because I was getting weird results consistently.

**Reply by JanIngoHaas** [reply]
Templates are applied. I face the same problem. Have you fixed it somehow?

**Reply by Madd0g** [reply]
my problem was that I wasn't sure if I should apply the template myself or not, in my own code.

I just stopped applying the template. Also tried passing mode=chat and mode=instruct but it just made things worse (at least in one case I remember with autogen).

**Reply by JanIngoHaas** [reply]
I don't want it to apply any template, that's the thing. It should properly chain the messages together using ChatML for example, but not do anything else. The "template" comes from my side.

**Reply by Madd0g** [reply]
sorry if I wasn't clear - the chatml wrapping is what I call "template". Maybe we're not talking about the same thing.

I was adding the chatml parts to the messages within my own code - this worked great with ollama+litellm because litellm just glued the user/assistant messages to each other with zero formatting. But then when I switched to oobabooga, it seemed like I didn't have to do it anymore because the api was doing it.

**Reply by DaOnlyOwner** [reply]
Gotcha. My problem was that somehow the "Chat template" from corresponding tab got applied (i.e. name1: XXX), but it now seems that the model did this and the template. Anyways, thanks for the help, I kinda solved it :)
